File: web/src/api/project.ts
import http from './axios'
import { TaskGroup } from './taskgroup'
import { User } from './user'

export interface Project {
  id?: number
  name?: string
  startAt?: string
  endAt?: string
  desc?: string
  ownerId?: number
  owner?: User
  taskGroups?: TaskGroup[]
  status?: number
}
const url = '/api/project'
export const addProject = (p: Project) => {
  return http({
    method: 'POST',
    url: url,
    data: p,
  }).then((res) => res.data)
}

export const lProjects = () => {
  return http({
    method: 'GET',
    url: url,
    params: {},
  }).then((res) => res.data)
}

export const getProjectDetail = (id: number) => {
  return http({
    method: 'GET',
    url: url + '/detail',
    params: { id: id },
  }).then((res) => res.data)
}

export const updateProject = (p: Project) => {
  return http({
    method: 'PUT',
    url: url,
    data: p,
  }).then((res) => res.data)
}

File: server/api/project.go
package api

import (
	"gopkg.in/guregu/null.v4"
)

const (
	StatusNotStart = 1
	StatusOn       = 2
	StatusDone     = 3
)

type Project struct {
	Id         uint         `json:"id" form:"id"`
	Name       string       `json:"name" form:"name"`
	StartAt    null.Time    `json:"startAt" form:"startAt"`
	EndAt      null.Time    `json:"endAt" form:"endAt"`
	Desc       string       `json:"desc" form:"desc"`
	Status     int          `json:"status" form:"status"`
	TaskGroups []*TaskGroup `json:"taskGroups" form:"taskGroups"`
	Owner      *User        `json:"owner" form:"owner"`
	OwnerId    uint         `json:"ownerId" form:"ownerId"`
	Color      string       `json:"color" form:"color"`
}

type ListProjectReq struct {
	ListOpt
	My bool `json:"my" form:"my"`
}
